pkg/controller/elasticsearch/client/error.go

```diff
@@ -0,0 +1,92 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+package client
+
+import (
+	"bytes"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+)
+
+// APIError is a non 2xx response from the Elasticsearch API
+type APIError struct {
+	Status        string
+	StatusCode    int
+	ErrorResponse ErrorResponse
+}
+
+// newAPIError converts an HTTP response into an API error, attempting to parse the body to include the details about the error.
+func newAPIError(response *http.Response) error {
+	defer response.Body.Close()
+	apiError := &APIError{
+		Status:     response.Status,
+		StatusCode: response.StatusCode,
+	}
+	// We may need to read the body multiple times, read the full body and store it as an array of bytes.
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		// We were not able to read the body, log this I/O error and return the API error with the status.
+		log.Error(err, "Cannot read error body")
+		return apiError
+	}
+	// Reset the response body to the original unread state. It allows a caller to read again the body if necessary,
+	// for example in the case of a 408.
+	response.Body = ioutil.NopCloser(bytes.NewBuffer(body))
```

Comment:
Do we read that body in case of 408? Or just check the status code?
The code flow here gets a bit tricky. I'm wondering if it may be simpler to always return either `response, nil` or `nil, err` for any http request we make. So that function that creates the `APIError` would read the response body, and then we don't care about that response body anymore, and there's no way to retrieve it (it is still stored in the error object if we need it).
Checking whether the api call we made returned a 408 would be a call to a helper function like `client.IsTimeout(err)`.

Reply:
> Do we read that body in case of 408? Or just check the status code?

In case of a 408, and if the path is `/_cluster/health`, then the error is later skipped, and `baseClient.request(..)` attempts to read again the body from the response.

I think I wanted to be conservative with this PR, keeping the responsibility of reading the "common" errors here, while delegating the more specific ones to the callers. This is the current flow.

Reply:
🤔 I tend to feel like this is more complicated than it should be, but I'm fine with keeping that PR small and focused on the problem it solves.